yt/yt/library/profiling/solomon/proxy.cpp

```diff
@@ -0,0 +1,410 @@
+{
+    std::vector<IEndpointProvider::TEndpoint> filteredEndpoints;
+    filteredEndpoints.reserve(endpoints.size());
+
+    for (const auto& endpoint : endpoints) {
+        // Skip non-selected instances if instance names are specified.
+        if (Instances && !Instances.contains(endpoint.Name)) {
+            continue;
+        }
+
+        // For all specified labels, skip instances that do not have the selected values of these labels (or any value).
+        bool matchesLabelFilter = true;
+        for (const auto& [label, acceptableValues] : LabelFilter) {
+            auto instanceLabelValueIt = endpoint.Labels.find(label);
+            if (instanceLabelValueIt == endpoint.Labels.end() || !acceptableValues.contains(instanceLabelValueIt->second)) {
+                matchesLabelFilter = false;
+                break;
+            }
+        }
+        if (!matchesLabelFilter) {
+            continue;
+        }
+
+        filteredEndpoints.push_back(endpoint);
+    }
+
+    return filteredEndpoints;
+}
+
+std::vector<TString> TSolomonProxy::CollectEndpoints(const TCgiParameters& params, int shard_index, int shard_count) const
+{
+    auto componentFilter = GetComponentFilter(params);
+    auto instanceFilter = GetInstanceFilter(params);
+
+    YT_LOG_DEBUG(
+        "Selecting endpoints (PublicComponentNames: %v, RequestedComponents: %v, RequestedInstances: %v, InstanceLabelFilter: %v)",
+        Config_->PublicComponentNames,
+        componentFilter.Components,
+        instanceFilter.Instances,
+        instanceFilter.LabelFilter);
+
+    int selectedComponents = 0;
+    int skippedByInternalFilter = 0;
+    int skippedByUserFilter = 0;
+
+    std::vector<TString> allEndpoints;
+
+    for (const auto& [componentName, endpointProvider] : ComponentNameToEndpointProvider_) {
+        YT_VERIFY(componentName == endpointProvider->GetComponentName());
+
+        if (Config_->PublicComponentNames && !Config_->PublicComponentNames->contains(componentName)) {
+            ++skippedByInternalFilter;
+            YT_LOG_DEBUG("Skipping non-public component (ComponentName: %v)", componentName);
+            continue;
+        }
+
+        if (!componentFilter.Match(endpointProvider)) {
+            ++skippedByUserFilter;
+            YT_LOG_DEBUG("Skipping non-requested component (ComponentName: %v)", componentName);
+            continue;
+        }
+
+        ++selectedComponents;
+
+        auto componentEndpoints = endpointProvider->GetEndpoints();
+        auto selectedEndpointCountBefore = allEndpoints.size();
+        for (const auto& endpoint : instanceFilter.Filter(componentEndpoints)) {
+            if (endpoint.GetFarmFingerprint() % shard_count == shard_index) {
+                allEndpoints.push_back(endpoint.Address);
+            }
+        }
+
+        YT_LOG_DEBUG(
+            "Selected component endpoints for pull (ComponentName: %v, SelectedEndpointCount: %v, TotalEndpointCount: %v)",
```

Comment:
I spent quite some time trying to figure out a non-ambiguous wording for the log parameters that you are proposing, because the order of label/shard filtering is not obvious.

I guess we can print all 4 values (shard & label filter, total, shard filter, label filter), even though it requires some extra computations.